mllib/src/main/scala/org/apache/spark/mllib/util/MLUtils.scala

```diff
@@ -267,6 +268,26 @@ object MLUtils {
   }
 
   /**
+   * :: Experimental ::
+   * Return a k element array of pairs of RDDs with the first element of each pair
+   * containing the training data, a complement of the validation data and the second
+   * element, the validation data, containing a unique 1/kth of the data. Where k=numFolds.
+   * The training and validation data are stratified by the key of the rdd, and the key
+   * ratios in the original data are maintained in each stratum of the train and validation
+   * data.
+   */
+  @Experimental
+  def kFoldStratified[K: ClassTag, V: ClassTag](rdd: RDD[(K, V)],
+      numFolds: Int,
+      seed: Int): Array[(RDD[(K, V)], RDD[(K, V)])] = {
+    val keys = rdd.keys.collect().distinct
```

Comment:
Best to swap the `distinct` and `collect` calls to avoid running out of memory on the driver.